### PR TITLE
refactor: migrate to pixi, modernize Python / sklearn / pandas floors

### PR DESCRIPTION
**Pull request by thomas-maschler**
# Modernize: pixi, Python 3.11+, sklearn 1.8+, pandas 2.2+

## Summary

Bundled modernization sweep — replaces poetry/Makefile with a pixi build, bumps the dep floors to versions that actually install in 2026, migrates to sklearn's new public APIs, fixes two latent bugs surfaced along the way, and swaps `black + flake8 + isort` for ruff. Verified across **9 CI matrix legs** (3 OSes × 3 Python versions) plus a `dev-glmnet` env that exercises the glmnet code path.

### Build system — pixi replaces poetry + Makefile + MANIFEST.in

- New `pyproject.toml` is PEP-621 `[project]` + `[tool.pixi.*]` + `[tool.hatch.*]`.
- Library deps live only in `[project.dependencies]` (pip-installable surface). `[tool.pixi.dependencies]` is just `python`.
- Single composite GH Action `.github/actions/setup-pixi` pins pixi version once for all 4 workflows.
- Tasks: `test`, `test-data`, `lint`, `format`, `format-check`, `build`, `build-docs`, `deploy-docs`, `localdocs`, `deploy`.

### Dep floors raised to match reality

| | before | after | reason |
|---|---|---|---|
| python | `>=3.9` | `>=3.11` | matches the floor that all current deps need |
| numpy | `>=2.0` | `>=2.0` | unchanged; was already accurate |
| pandas | `>=1.0.3,<3.0` | `>=2.2.2` | floor was fictional given `numpy>=2.0`; cap removed |
| scikit-learn | `>=1.2,<1.6` | `>=1.8` | required for `__sklearn_tags__` + the `l1_ratio` API |
| matplotlib | `>=3.7` | `>=3.10` | for `boxplot(tick_labels=, orientation=)` |

### sklearn 1.8+ API migration

- `SDMMixin` now uses `__sklearn_tags__()` (1.6+) instead of the deprecated `_more_tags()` + `_estimator_type` class attribute. Inherits `ClassifierMixin`; reorders bases so the override actually wins MRO lookup.
- `LogisticRegression(penalty="l1", solver="liblinear")` → `LogisticRegression(l1_ratio=1.0, solver="liblinear")`. **Empirically verified bit-identical**: max coefficient delta = 0.0, same iteration count on the ariolimax sample. The old API would error at construction time in sklearn 1.10.
- `partial_dependence_scores` now casts integer-dtype features to float64 before calling sklearn (1.7+ rejects integer feature columns); also NaN-pads per-feature grids when sklearn collapses to fewer-than-`n_bins` unique values.
- `SDMMixin.classes_` is now `np.array([0, 1])` instead of `[0, 1]` — sklearn 1.9+ uses boolean masking against `classes_` and silently failed on the Python list form.

### Two latent bugs caught and fixed

1. **`features.left_hinge` / `right_hinge`** called `np.divide(num, denom, where=valid)` without an `out=` array. The `~valid` cells contained uninitialized memory; the subsequent `np.clip(0, 1)` doesn't clip NaN/Inf — so any feature with a zero training range (`mn == mx`) could leak NaN/Inf into the feature matrix. Fixed with `out=np.zeros(...)` so unfilled cells default to 0.
2. **`MaxentModel.fit`** applied `preprocessor.transform(x)` once, then called `self.predict(x[y==0])` internally to compute `alpha_` / `entropy_`. `predict()` re-applied the preprocessor → background statistics were computed on **doubly-transformed** data. For PCA this is meaningful corruption of `entropy_`, which feeds the logistic/cloglog outputs. Fixed by passing the un-preprocessed `x` to the internal predict.

### Lint: ruff replaces black + flake8 + isort

- `[tool.ruff]` covers formatting + linting. Line length 120 (matches the existing repo style).
- Pre-commit hook uses `language: system` so it runs the env-installed ruff binary — single source of truth instead of duplicate pinned versions.
- Initial `ruff check --fix` + `--unsafe-fixes` plus manual cleanup: 68 issues fixed (dead `typing.List/Dict/Tuple` imports, `Optional` → `| None`, `zip(..., strict=)`, `warnings.warn(..., stacklevel=2)`, mutable default args, `yield-from` collapses). Final state: `All checks passed!`

### CI matrix expanded

```yaml
matrix:
  os: [ubuntu-latest, macos-latest, windows-latest]
  environment: [dev, dev-py312, dev-py313]
```

Down from two separate workflows (`run-pytest` + `os-tests`) to one, with Python and OS axes combined. Per-Python pixi envs (`dev` = 3.11, `dev-py312`, `dev-py313`) let the matrix actually exercise multiple Python versions through pixi.

### glmnet — broken pip path, working conda path

The `pip install elapid[glmnet]` extra has been **broken since 2020** when the upstream `glmnet` package went dormant — its `setup.py` predates PEP 517 and no current installer can build it. Empirically reproduced.

Solution:

- Drop the `[project.optional-dependencies] glmnet` extra (the broken promise).
- Add a `dev-glmnet` pixi environment that pulls the **conda-forge** glmnet binary, with `[tool.pixi.feature.glmnet] platforms = ["linux-64", "osx-64", "osx-arm64"]` (conda-forge has no Windows build). Pin `setuptools<80` because glmnet imports `pkg_resources` which setuptools 80+ dropped.
- `docs/install.md` rewritten to explain the install paths honestly.

`pixi run -e dev-glmnet test` exercises the glmnet branch — coverage jumps from 88% → **90%**.

### docs `--strict` build now passes

Eight `griffe` warnings under `mkdocs --strict` fixed:

- 5× `**kwargs` parameters annotated with `: Any` (geo, models, stats).
- `zonal_stats` Args section split out — was using a comma-separated docstring grouping `mean, min, max, count, sum, stdv, skew, kurtosis, mode:` that griffe parsed as a single fictitious parameter.
- `format_band_labels` got its missing `-> list[str]` return annotation.

### Other touch-ups

- `[tool.pytest.ini_options]` holds the default test addopts; CI passes extra flags (`--cov-report=xml:coverage.xml`) as trailing args. Dropped the redundant `test-ci` task.
- `test_download_sample_data` uses `tempfile.TemporaryDirectory()` for guaranteed cleanup (the old version leaked a `tmp-download/` folder in the repo root on any earlier assertion failure — found one such artifact in the working tree).
- `percentile_range: tuple[float, float] = [2.5, 97.5]` → `(2.5, 97.5)`; same for `categorical_features` default in `partial_dependence_scores`.
- 7 new regression tests in `tests/test_models.py` and `tests/test_features.py` covering: `is_classifier()` for all three model classes, the `__sklearn_tags__` shape, the `classes_` ndarray invariant, the int-dtype `partial_dependence` path, the `_cast_integer_features_to_float` helper (mutation + passthrough), and the hinge-memory zero-range edge case.

## Test plan

- [x] `pixi run -e dev test` — 70 passed on Python 3.11.15
- [x] `pixi run -e dev-py312 test` — 70 passed on Python 3.12.13
- [x] `pixi run -e dev-py313 test` — 70 passed on Python 3.13.13
- [x] `pixi run -e dev-glmnet test` — 70 passed (90% coverage with glmnet path active)
- [x] `pixi run -e dev lint` + `format-check` — clean
- [x] `pixi run -e dev pytest -W error` — passes (zero warnings remaining; matplotlib boxplot/penalty deprecations also fixed)
- [x] `pixi run -e docs build-docs` (mkdocs --strict) — clean, 0 griffe warnings
- [x] `pixi run -e build build` — produces sdist + wheel with `elapid/data/*.{gpkg,csv.gz}` packaged

## Breaking changes for downstream consumers

- `pip install elapid[glmnet]` no longer exists. Use `conda install -c conda-forge elapid glmnet` instead (the path that's actually been working). For elapid contributors, `pixi run -e dev-glmnet ...` provides the same path locally.
- Python 3.9 and 3.10 are no longer supported (floor raised to 3.11).

### CODE REVIEW

File: elapid/features.py [hunk 1]
```diff
@@ -588,7 +588,8 @@ def left_hinge(x: ArrayLike, mn: float, mx: float) -> np.ndarray:
     """
     rng = repeat_array(mx, mn.shape[-1], axis=1) - mn
     valid = rng > 0
-    hinge = np.clip(np.divide((x - mn), rng, where=valid), 0, 1)
+    out = np.zeros_like(x - mn, dtype=np.float64)
+    hinge = np.clip(np.divide((x - mn), rng, where=valid, out=out), 0, 1)
```
Comment (thomas-maschler) [inline elapid/features.py:592]: Pre-existing bug fix. np.divide(num, denom, where=valid) without an out= array returned a freshly-allocated array whose ~valid cells contained uninitialized memory. The subsequent np.clip(0, 1) does not clip NaN/Inf, so any feature with a zero training range (mn == mx) could leak NaN/Inf into the feature matrix. Passing out=np.zeros_like(...) makes the skipped cells deterministically 0, which is the semantically correct value for "zero-width range carries no information."

Caught by the new test_hinges_no_uninitialized_memory_when_range_is_zero regression test in tests/test_features.py.

File: elapid/models.py [hunk 2]
```diff
@@ -165,25 +176,28 @@ def partial_dependence_scores(
         Returns:
             bins, mean, stdv: the binned feature values and the mean/stdv of responses.
         """
+        x = _cast_integer_features_to_float(x)
```
Comment (thomas-maschler) [inline elapid/models.py:179]: sklearn 1.7 made integer-dtype feature columns a hard error in partial_dependence (previously a silent type-promotion). The cast happens once here on a non-mutating copy so the rest of the function can pass the same x to sklearn for every feature index. See _cast_integer_features_to_float at the bottom of the file for the helper's docstring.

File: elapid/models.py [hunk 3]
```diff
@@ -165,25 +176,28 @@ def partial_dependence_scores(
...
+
         ncols = x.shape[1]
-        mean = np.zeros((ncols, n_bins))
-        stdv = np.zeros_like(mean)
-        bins = np.zeros_like(mean)
+        mean = np.full((ncols, n_bins), np.nan)
+        stdv = np.full((ncols, n_bins), np.nan)
+        bins = np.full((ncols, n_bins), np.nan)
```
Comment (thomas-maschler) [inline elapid/models.py:185]: NaN-padded rather than zero-padded because sklearn's partial_dependence collapses the grid to a feature's unique values when fewer than grid_resolution exist (e.g. an integer-discrete column cast to float). When the grid is shorter than n_bins, the trailing slots stay NaN and the partial_dependence_plot skips them naturally (matplotlib doesn't draw NaN segments). The fixed-shape (ncols, n_bins) return contract is preserved.

File: elapid/models.py [hunk 4]
```diff
@@ -369,6 +386,7 @@ def fit(
         y = format_occurrence_data(y)
 
         # apply preprocessing
+        x_raw = x
```
Comment (thomas-maschler) [inline elapid/models.py:395]: Pre-existing bug fix. The original code applied the preprocessor in-place on x, then later called self.predict(x[y==0]) to compute alpha_ / entropy_. predict() re-applies the preprocessor — so the background statistics were computed on doubly-transformed data. For PCA this is meaningful corruption of entropy_, which feeds the logistic and cloglog outputs. Keeping a reference to the un-preprocessed x and passing that to the internal predict call is the minimal fix.

Caught and guarded by the strengthened test_preprocessor in tests/test_models.py, which now elevates the "X does not have valid feature names" warning to an error.

File: elapid/models.py [hunk 5]
```diff
@@ -443,7 +461,7 @@ def fit(
         # apply maxent-specific transformations
         class_transform = self.get_params()["transform"]
         self.set_params(transform="raw")
-        raw = self.predict(x[y == 0])
+        raw = self.predict(x_raw[y == 0])
```
Comment (thomas-maschler) [inline elapid/models.py:470]: Companion to the preprocessor fix above. x_raw (the original, un-preprocessed input) is passed here on purpose — predict() will apply the preprocessor exactly once internally, which is what the alpha/entropy calculation requires.

File: elapid/models.py [hunk 6]
```diff
@@ -560,20 +578,20 @@ def initialize_sklearn_model(self, C: float, fit_intercept: bool = True) -> None
         self.estimator = LogisticRegression(
             C=C,
             fit_intercept=fit_intercept,
-            penalty="l1",
+            l1_ratio=1.0,
             solver="liblinear",
             tol=self.convergence_tolerance,
-            max_iter=self.n_lambdas,
+            max_iter=1000,
```
Comment (thomas-maschler) [inline elapid/models.py]: Previously this was max_iter=self.n_lambdas, which conflated two unrelated knobs: n_lambdas is the regularization-path length used by the glmnet path; liblinear's coordinate-descent solver has no notion of a regularization path. The default n_lambdas=100 is too few iterations to converge at the tight MaxentConfig.tolerance=1e-7 default, which is what was producing the ConvergenceWarning: Liblinear failed to converge you might have seen pre-PR. 1000 is generous for the size of feature matrices elapid typically produces.

File: elapid/features.py (hunk 1 above)
Comment (earth-chris) [reply elapid/features.py:592]: great catch

File: elapid/models.py (hunk 4 above)
Comment (earth-chris) [reply elapid/models.py:395]: great catch, i remember this being an issue but didn't know how to debug.

File: elapid/models.py (hunk 6 above)
Comment (earth-chris) [reply elapid/models.py]: nice catch again. I might prefer adding this value to the `config.MaxentConfig` instead of specifying a magic number, but am also ok with passing this as-is

**Review by earth-chris — APPROVED** [review]
